fedimint-testing/src/gateway.rs

```diff
@@ -113,44 +116,50 @@ impl GatewayTest {
         )
         .await
         .expect("Failed to create gateway");
-        gateway
-            .clone()
-            .run()
-            .await
-            .expect("Failed to start gateway");
-
-        // Wait for the gateway to be in the running state
-        let mut gateway_state_iterations = 0;
-        loop {
-            if let GatewayState::Running {
-                lnrpc: _,
-                lightning_public_key: _,
-                lightning_alias: _,
-            } = gateway.state.read().await.clone()
-            {
-                break;
-            }
 
-            if gateway_state_iterations >= 30 {
-                panic!("Gateway did not start running after 10 attempts");
-            }
+        let mut task_group = TaskGroup::new();
+        let gateway_run = gateway.clone();
+        task_group
+            .spawn("Gateway Run", |_handle| async move {
+                gateway_run.run().await.expect("Failed to start gateway");
+            })
+            .await;
 
-            gateway_state_iterations += 1;
-            sleep(Duration::from_secs(1)).await;
-        }
+        // Wait for the gateway to be in the configuring or running state
+        GatewayTest::wait_for_gateway_state(gateway.clone(), |gw_state| {
+            matches!(gw_state, GatewayState::Configuring)
+                || matches!(gw_state, GatewayState::Running { .. })
+        })
+        .await;
 
         let listening_addr = lightning.listening_address();
         let info = lightning.info().await.unwrap();
 
         Self {
-            password,
             api: address,
             _config_dir,
             gateway,
             node_pub_key: PublicKey::from_slice(info.pub_key.as_slice()).unwrap(),
             listening_addr,
         }
     }
+
+    pub async fn wait_for_gateway_state(gateway: Gateway, func: impl Fn(GatewayState) -> bool) {
+        let mut gateway_state_iterations = 0;
+        loop {
+            let gw_state = gateway.state.read().await.clone();
+            if func(gw_state) {
+                break;
+            }
+
+            if gateway_state_iterations >= 30 {
+                panic!("Gateway did not start running after 10 attempts");
```

Comment:
Is it always the "started" state that is awaited? (just because the fn is now parametrized with a predicate)

Reply:
This is used in two spots. One spot waits for `Configuring` or `Running` (to test the case where the password has not been set at all) and the other case only waits for `Running` since it has already set the password.